components/brave_wallet/browser/zcash/rust/orchard_shard_tree_impl.cc

```diff
@@ -0,0 +1,358 @@
+  }
+  return ShardStoreStatusCode::Ok;
+}
+
+ShardStoreStatusCode ShardTreeDelegate::RemoveCheckpoint(
+    uint32_t checkpoint_id) const {
+  auto result = delegate_->RemoveCheckpoint(checkpoint_id);
+  if (!result.has_value()) {
+    return ShardStoreStatusCode::Error;
+  } else if (!result.value()) {
+    return ShardStoreStatusCode::None;
+  }
+  return ShardStoreStatusCode::Ok;
+}
+
+ShardStoreStatusCode ShardTreeDelegate::TruncateCheckpoint(
+    uint32_t checkpoint_id) const {
+  auto result = delegate_->TruncateCheckpoints(checkpoint_id);
+  if (!result.has_value()) {
+    return ShardStoreStatusCode::Error;
+  } else if (!result.value()) {
+    return ShardStoreStatusCode::None;
+  }
+  return ShardStoreStatusCode::Ok;
+}
+
+ShardStoreStatusCode ShardTreeDelegate::GetCheckpoints(
+    size_t limit,
+    ::rust::Vec<ShardTreeCheckpointBundle>& into) const {
+  auto checkpoints = delegate_->GetCheckpoints(limit);
+  if (!checkpoints.has_value()) {
+    return ShardStoreStatusCode::Error;
+  }
+  if (checkpoints->empty()) {
+    return ShardStoreStatusCode::None;
+  }
+  for (const auto& checkpoint : checkpoints.value()) {
+    into.push_back(From(checkpoint));
+  }
+  return ShardStoreStatusCode::Ok;
+}
+
+bool OrchardShardTreeImpl::ApplyScanResults(
+    std::unique_ptr<OrchardDecodedBlocksBundle> commitments) {
+  auto* bundle_impl =
+      static_cast<OrchardDecodedBlocksBundleImpl*>(commitments.get());
+  return orcard_shard_tree_->insert_commitments(bundle_impl->GetDecodeBundle());
+}
+
+base::expected<OrchardNoteWitness, std::string>
+OrchardShardTreeImpl::CalculateWitness(uint32_t note_commitment_tree_position,
+                                       uint32_t checkpoint) {
+  auto result = orcard_shard_tree_->calculate_witness(
+      note_commitment_tree_position, checkpoint);
+  if (!result->is_ok()) {
+    return base::unexpected(result->error_message().c_str());
+  }
+
+  auto value = result->unwrap();
+
+  OrchardNoteWitness witness;
+  witness.position = note_commitment_tree_position;
+  for (size_t i = 0; i < value->size(); i++) {
+    witness.merkle_path.push_back(value->item(i));
+  }
+
+  return witness;
+}
+
+bool OrchardShardTreeImpl::TruncateToCheckpoint(uint32_t checkpoint_id) {
+  return orcard_shard_tree_->truncate(checkpoint_id);
+}
+
+OrchardShardTreeImpl::OrchardShardTreeImpl(
+    rust::Box<OrchardShardTreeBundle> orcard_shard_tree)
+    : orcard_shard_tree_(std::move(orcard_shard_tree)) {}
+
+OrchardShardTreeImpl::~OrchardShardTreeImpl() {}
+
+// static
+std::unique_ptr<OrchardShardTree> OrchardShardTree::Create(
+    std::unique_ptr<::brave_wallet::OrchardShardTreeDelegate> delegate) {
+  auto shard_tree_result = ::brave_wallet::orchard::create_shard_tree(
+      std::make_unique<ShardTreeDelegate>(std::move(delegate)));
+  if (!shard_tree_result->is_ok()) {
+    return nullptr;
+  }
+  return base::WrapUnique<OrchardShardTree>(
+      new OrchardShardTreeImpl(shard_tree_result->unwrap()));
+}
```

Comment:
make_unique

Reply:
Constructor is private like in other classes from this dir